ql/pricingengines/forward/mcforwardeuropeanhestonengine.hpp

```diff
@@ -0,0 +1,249 @@
+/* -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*
+ This file is part of QuantLib, a free-software/open-source library
+ for financial quantitative analysts and developers - http://quantlib.org/
+ QuantLib is free software: you can redistribute it and/or modify it
+ under the terms of the QuantLib license.  You should have received a
+ copy of the license along with this program; if not, please email
+ <quantlib-dev@lists.sf.net>. The license is also available online at
+ <http://quantlib.org/license.shtml>.
+ This program is distributed in the hope that it will be useful, but WITHOUT
+ ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ FOR A PARTICULAR PURPOSE.  See the license for more details.
+*/
+
+/*! \file mcforwardvanillahestonengine.hpp
+    \brief Monte Carlo engine for forward-starting strike-reset vanilla options using Heston-like process
+*/
+
+#ifndef quantlib_mc_forward_european_heston_engine_hpp
+#define quantlib_mc_forward_european_heston_engine_hpp
+
+#include <ql/pricingengines/forward/mcforwardvanillaengine.hpp>
+#include <ql/processes/hestonprocess.hpp>
+
+namespace QuantLib {
+
+    /*! \ingroup forwardengines
+        \test
+        - Heston MC prices for a flat Heston process are
+          compared to analytical BS prices with the same
+          volatility for a range of moneynesses
+        - Heston MC prices for a forward-starting option
+          resetting at  t=0 are compared to semi-analytical
+          Heston prices for a range of moneynesses
+    */
+    template <class RNG = PseudoRandom,
+              class S = Statistics, class P = HestonProcess>
```

Comment:
Is the template to arrange both Heston and Bates?

Reply:
Bates inherits from Heston so I think both should be ok? Although I haven't tested creating a Bates version so far. I tried to follow the current implementation of `mceuropeanhestonengine.hpp` as much as possible.

Reply:
You're right, somehow I missed that `MCEuropeanHestonEngine` is also a template in the same way 😄